src/bandit-evaluator.ts

```diff
@@ -0,0 +1,178 @@
+          coefficients.actionNumericCoefficients,
+          actionAttributes,
+        );
+        score += this.scoreCategoricalAttributes(
+          coefficients.actionCategoricalCoefficients,
+          actionAttributes,
+        );
+      }
+      actionScores[actionKey] = score;
+    });
+    return actionScores;
+  }
+
+  private scoreNumericAttributes(
+    coefficients: BanditNumericAttributeCoefficients[],
+    attributes: Attributes,
+  ): number {
+    return coefficients.reduce((score, numericCoefficients) => {
+      const attributeValue = attributes[numericCoefficients.attributeKey];
+      if (typeof attributeValue === 'number' && isFinite(attributeValue)) {
+        score += attributeValue * numericCoefficients.coefficient;
+      } else {
+        score += numericCoefficients.missingValueCoefficient;
+      }
+      return score;
+    }, 0);
+  }
+
+  private scoreCategoricalAttributes(
+    coefficients: BanditCategoricalAttributeCoefficients[],
+    attributes: Attributes,
+  ): number {
+    return coefficients.reduce((score, attributeCoefficients) => {
+      const attributeValue = attributes[attributeCoefficients.attributeKey]?.toString();
+      const applicableCoefficient =
+        attributeValue && attributeCoefficients.valueCoefficients[attributeValue];
+
+      score +=
+        typeof applicableCoefficient === 'number'
+          ? applicableCoefficient
+          : attributeCoefficients.missingValueCoefficient;
+
+      return score;
+    }, 0);
+  }
+
+  private weighActions(
+    actionScores: Record<string, number>,
+    gamma: number,
+    actionProbabilityFloor: number,
+  ) {
+    const actionWeights: Record<string, number> = {};
+    const actionScoreEntries = Object.entries(actionScores);
+
+    if (!actionScoreEntries.length) {
+      return actionWeights;
+    }
+
+    // First find the action with the highest score
+    let currTopScore: number | null = null;
+    let currTopAction: string | null = null;
+    actionScoreEntries.forEach(([actionKey, actionScore]) => {
+      if (currTopScore === null || actionScore > currTopScore) {
+        currTopScore = actionScore;
+        currTopAction = actionKey;
+      }
+    });
+
+    if (currTopScore === null || currTopAction === null) {
+      // Appease typescript with this check and extra variables
+      throw new Error('Unable to find top score');
+    }
+    const topScore: number = currTopScore;
+    const topAction: string = currTopAction;
+
+    // Then weigh every action but the top one
+    const numActions = actionScoreEntries.length;
+    const minimumWeight = actionProbabilityFloor / numActions;
+    let cumulativeWeight = 0;
+
+    actionScoreEntries.forEach(([actionKey, actionScore]) => {
+      if (actionKey === topAction) {
+        // We weigh the top action later
+        return;
+      }
+      const weight = 1 / (numActions + gamma * (topScore - actionScore));
+      const boundedWeight = Math.max(weight, minimumWeight);
```

Comment:
Note: in the Java SDK I opted to have an extra step to round the weight to four decimal places ([source link](https://github.com/Eppo-exp/java-server-sdk/blob/main/src/main/java/com/eppo/sdk/helpers/bandit/FalconBanditModel.java#L104)), to account for the possibility of different languages treating fractional numbers differently. We do not do this in the Python SDK ([source link](https://github.com/Eppo-exp/python-sdk/blob/main/eppo_client/bandit.py#L172)), which I think is OK as we've since confirmed all our SDK languages use IEEE 754 representation. Thus, I don't do it here either.